src/particles.cpp

```diff
@@ -291,4 +292,7 @@ void ParticleParameters::deSerialize(std::istream &is, u16 protocol_ver)
 		return;
 	jitter.deSerialize(is);
 	bounce.deSerialize(is);
+	if (is.eof())
+		return;
+	texture.deSerialize(is, protocol_ver, true);
```

Comment:
`is.eof()` only returns true after you have read over the end of the stream, so this probably won't work as expected.

That's why there is e.g. this code:

https://github.com/minetest/minetest/blob/4859cf44ce3259d7b7fbdcf37055c68cc9363805/src/nodedef.cpp#L655-L684

Reply:
Should be fixed in https://github.com/minetest/minetest/pull/14353/commits/bb6b43d9b5364f6eddebfd61751275219e28140e.

I think the `SerializationError` hack is very bad. Even if we decide to (ab)use error handling for this, that should be separate from serialization errors. Currently we're silencing "real" serialization errors in backwards compatibility code.

I find `streamEndsBeforeParam` to be a better hack. In general I think that the `deSerialize` methods should not be populating a default-initialized object, but rather returning the object they deserialize. Then you could throw without fear of ending up in an inconsistent state.

We could have our own stream subclass which supports proper "ahead of time" EOF. I think most of our streams are backed by buffers with known length anyways (coming from packets).